mitodl/unified-ecommerce
# Adds payment API and cart test mule app

### What are the relevant tickets?

https://github.com/mitodl/hq/issues/3166
https://github.com/mitodl/hq/issues/3158

### Description (What does it do?)

Adds the payment API, allowing users to check out via API calls.

Also adds a cart test mule app, which is a very simple frontend that allows users to purchase an item. (Apart from testing, this is a placeholder so that there's a defined space for the actual frontend that this will get; it also holds the interstitial page for redirecting the user into the hosted payment provider.) 

There are also a handful of minor fixes to support the above:
- There is a `docker-compose` change that allows you to specify the port for Postgres by setting `POSTGRES_PORT` in the `.env`. It will default to `5432` - if you have other services, you may want to change that, though.
- The `generate_test_data` command was updated to properly work with revisions for Product objects (it didn't before), and to ensure that the products it creates have prices < $1000 (so you can actually buy them - CyberSource won't authorize a test payment greater than $999). 

### Screenshots (if appropriate):

![Screenshot 2024-04-10 at 10 39 59 AM](https://github.com/mitodl/unified-ecommerce/assets/945611/7dd252f0-2987-4c1f-b985-6e25c3749224)
![Screenshot 2024-04-10 at 10 39 50 AM](https://github.com/mitodl/unified-ecommerce/assets/945611/74552a61-a343-407e-af78-e2e6257ebc2a)
![Screenshot 2024-04-10 at 10 46 15 AM](https://github.com/mitodl/unified-ecommerce/assets/945611/f0034efe-ef93-4e7a-9fac-10791aa11125)

### How can this be tested?

You will need a working environment set up. This includes Keycloak, and either APISIX or Traefik. See the `README.md` for more details. You should also run `generate_test_data` if you don't already have a system and products set up. 

Automated tests should pass.

Navigating to the homepage - http://ue.odl.local (optionally with a port if you're using APISIX) - should get you a page now, rather than a 404. On that screen, you should be able to choose a product from the dropdown list, and then either add it to cart or add to cart and check out. If you add to cart, you can use the Check Out button to check out and the Clear Cart button should clear your cart successfully. 

Upon checkout, you should see an interstitial page that says "Redirecting to payment processor". The fields for the payload that will be sent to CyberSource are visible here as well as a "Redirect to CyberSource" button at the bottom. You can use this to verify information if you like; the page will automatically redirect in 5 seconds (there's a countdown timer in the button) or you can click the button to skip the timer. Once the redirect happens, you should be sent to CyberSource as per usual and completing the payment workflow there should work as expected. 

Upon successful completion, your cart should be cleared. You should also be able to see that the status of your order has changed to Fulfilled and a Transaction line should be created. There are no notifications on the cart screen at this time (and there likely won't be as this is far from the actual cart interface). No post-sale events will be triggered at this point either - there is future code that depends on this PR that will add that in. 

Cancelling the transaction in CyberSource should work as expected and your cart should remain intact.

Note that unlike other integrations the cart here _does_ support multiple items, so you should be able to check out with an arbitrary number of line items. The items internally link to the system they belong to and (once we have the post-sale events) will trigger actions in those specific systems once payment is complete.

### Additional Context

We'll hide the interstitial form and remove the delay later - it's a bit useful for testing now. (Probably behind a config setting, so it can be toggled back on if necessary.)

The cart app has had no professional styling and does not conform to our standards (for one, it's not a React app). It's solely to have an interface to use for testing. We'll add a real frontend later on but this is not the PR for it. (So, unless there's a dire need to add a feature to it, expanding this specific frontend interface is unlikely to happen.)

Post-sale events don't happen right now - there's a future PR coming up for that.
